File: Flyweight/c++/Flyweight.cpp
#include <iostream>
#include <string>
#include <unordered_map>

struct Shared
{
    std::string company;
    std::string position;

    Shared(const std::string company, const std::string position)
    : company(company), position(position) {}
};

struct Unique
{
    std::string passport;
    std::string name;

    Unique(const std::string name, const std::string passport)
    : passport(passport), name(name) {}
};

class Flyweight
{
    private:
        Shared* shared;

    public:
        Flyweight(const Shared* shared) : shared(new Shared(*shared)) {}
        Flyweight(const Flyweight& other) : shared(new Shared(*other.shared)) {}
        ~Flyweight();
        void process(const Unique& unique) const;
};

class FlyweightFactory
{
    private:
        std::unordered_map<std::string, Flyweight> flyweights;
        std::string GetKey(const Shared& shared) const;

    public:
        FlyweightFactory(std::initializer_list<Shared> shareds);
        Flyweight GetFlyweight(const Shared& shared);
        void listFlyweights() const;
};

void addSpecialistDatabase(FlyweightFactory& flyweightFactory, const std::string company,
                            const std::string position, const std::string name, const std::string passport);

int main()
{
    FlyweightFactory* flyweightFactory = new FlyweightFactory({
        {"Microsoft", "manager"},
        {"Google", "manager"},
        {"Google", "developer"},
        {"Apple", "developer"}
    });

    flyweightFactory->listFlyweights();

    addSpecialistDatabase(*flyweightFactory,
                           "Google",
                           "developer",
                           "George",
                           "12345");

    addSpecialistDatabase(*flyweightFactory,
                           "Apple",
                           "manager",
                           "Mark",
                           "67890");

    flyweightFactory->listFlyweights();

    return 0;
}

Flyweight::~Flyweight()
{
    delete shared;
}

void Flyweight::process(const Unique &unique) const
{
    std::cout << "New data: shared - " << shared->company << "_" <<
    shared->position << " unique - " << unique.name << "_" << unique.passport << std::endl;
}

std::string FlyweightFactory::GetKey(const Shared &shared) const
{
    return shared.company + "_" + shared.position;
}

FlyweightFactory::FlyweightFactory(std::initializer_list<Shared> shareds)
{
    for (const Shared& shared : shareds)
    {
        this->flyweights.insert(std::make_pair<std::string, Flyweight>(this->GetKey(shared), Flyweight(& shared)));
    }
}

Flyweight FlyweightFactory::GetFlyweight(const Shared &shared)
{
    std::string key = this->GetKey(shared);
    if (this->flyweights.find(key) == this->flyweights.end())
    {
        std::cout << "Flyweight factory: object not found " << key << std::endl;
        this->flyweights.insert(std::make_pair(key, Flyweight(&shared)));
    }
    else
    {
        std::cout << "Flyweight factory: object found " << key << std::endl;
    }
    return this->flyweights.at(key);
}

void FlyweightFactory::listFlyweights() const
{
    size_t count = this->flyweights.size();
    std::cout << "\nFlyweight factory: count = " << count << std::endl;
    for (std::pair<std::string, Flyweight> pair : this->flyweights)
        std::cout << pair.first << std::endl;
}

void addSpecialistDatabase(FlyweightFactory& flyweightFactory, const std::string company,
                            const std::string position, const std::string name, const std::string passport)
{
    std::cout << std::endl;
    const Flyweight& flyweight = flyweightFactory.GetFlyweight({ company, position });
    flyweight.process({ name, passport });
}

File: Flyweight/python/Flyweight.py
from typing import List
from typing import Dict


class Shared:

    def __init__(self, company: str, position: str):
        self.__company = company
        self.__position = position

    @property
    def company(self):
        return self.__company

    @property
    def position(self):
        return self.__position


class Unique:

    def __init__(self, name: str, passport: str):
        self.__name = name
        self.__passport = passport

    @property
    def name(self):
        return self.__name

    @property
    def passport(self):
        return self.__passport


class Flyweight:

    def __init__(self, shared: Shared):
        self.__shared = shared

    def process(self, unique: Unique):
        print("New data: shared - {}_{} unique - {}_{}".format(self.__shared.company, self.__shared.position, unique.name, unique.passport))

    def get_data(self) -> str:
        return self.__shared.company + '_' + self.__shared.position


class FlyweightFactory:

    def get_key(self, shared: Shared):
        return '{}_{}'.format(shared.company, shared.position)

    def __init__(self, shareds: List[Shared]):
        self.__flyweights: Dict[str, Flyweight] = {}
        for shared in shareds:
            self.__flyweights[self.get_key(shared)] = Flyweight(shared)

    def get_flyweight(self, shared: Shared):
        key: str = self.get_key(shared)
        if self.__flyweights.get(key) is None:
            print("Flyweight factory: object not found " + key)
            self.__flyweights[key] = Flyweight(shared)
        else:
            print("Flyweight factory: object found " + key)
        return self.__flyweights[key]

    def list_flyweights(self):
        count: int = len(self.__flyweights)
        print("\nFlyweight factory: count = {}".format(count))
        for pair in self.__flyweights.values():
            print(pair.get_data())


def add_specialist_database(flyweights_factory: FlyweightFactory,
                            company: str, position: str,
                            name:str, passport: str):
    print()
    flyweight = flyweights_factory.get_flyweight(Shared(company, position))
    flyweight.process(Unique(name, passport))


if __name__ == '__main__':

    shared_list: List[Shared] = (
        Shared("Microsoft", "manager"),
        Shared("Google", "manager"),
        Shared("Google", "developer"),
        Shared("Apple", "developer")
    )

    flyweight_factory = FlyweightFactory(shared_list)
    flyweight_factory.list_flyweights()

    add_specialist_database(flyweight_factory,
                            "Google",
                            "developer",
                            "George",
                            "12345")

    add_specialist_database(flyweight_factory,
                            "Apple",
                            "manager",
                            "Mark",
                            "67890")

    flyweight_factory.list_flyweights()